src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs

```diff
@@ -102,16 +109,54 @@ public async Task SpecializeAsync()
         {
             _logger.LogInformation("Starting language worker channel specialization");
             _workerRuntime = _environment.GetEnvironmentVariable(LanguageWorkerConstants.FunctionWorkerRuntimeSettingName);
+
             ILanguageWorkerChannel languageWorkerChannel = await GetChannelAsync(_workerRuntime);
             if (_workerRuntime != null && languageWorkerChannel != null)
             {
-                _logger.LogInformation("Loading environment variables for runtime: {runtime}", _workerRuntime);
-                await languageWorkerChannel.SendFunctionEnvironmentReloadRequest();
+                // Don't use placeholder process if app is not read-only or important environment config has changed
+                // TODO: Allow non read-only apps to use placeholder process - https://github.com/Azure/azure-functions-host/issues/4534
+                if (UsePlaceholderProcess(_workerRuntime))
+                {
+                    _logger.LogInformation("Loading environment variables for runtime: {runtime}", _workerRuntime);
+                    await languageWorkerChannel.SendFunctionEnvironmentReloadRequest();
+                }
+                else
+                {
+                    // The process spun up in placeholder mode is incompatible and we should shut it down
+                    _workerChannels.TryRemove(_workerRuntime, out Dictionary<string, TaskCompletionSource<ILanguageWorkerChannel>> languageWorkerChannels);
+                    await ShutdownChannelIfExistsAsync(_workerRuntime, languageWorkerChannel.Id);
+                }
             }
             _shutdownStandbyWorkerChannels();
             _logger.LogDebug("Completed language worker channel specialization");
         }
 
+        internal bool UsePlaceholderProcess(string selectedRuntime)
+        {
+            // App must be readonly
+            if (!_environment.FileSystemIsReadOnly())
+            {
+                return false;
+            }
+
+            // App must have specific environment configurations match
+            if (_initialEnvironmentConfig.TryGetValue(selectedRuntime, out Dictionary<string, string> environmentConfig))
+            {
+                foreach (string settingKey in environmentConfig.Keys)
```

Comment:
>foreach (string settingKey in environmentConfig.Keys) [](start = 16, length = 53)

If you're looping through the keys anyway, does this need to be a dictionary? Can it just be a List<> instead?